server.js

```diff
@@ -7,22 +7,145 @@ const mongoUrl = process.env.MONGO_URL || "mongodb://localhost/happyThoughts"
 mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true })
 mongoose.Promise = Promise
 
+
+const Thought = mongoose.model('Thought', {
+  message: {
+    type: String,
+    required: true,
+    minlength: 5,
+    maxlength: 140
+  },
+  hearts: {
+    type: Number,
+    default: 0
+  },
+  createdAt: {
+    type: Date,
+    default: Date.now
+  },
+  name: {
+    type: String,
+    required: false,
+    default: "anonymous",
+    maxlength: 50
+  },
+  tag: {
+    type: String,
+  }
+})
+
+//SEEDING FOR ADDING NEW DATA. When do I need to use/do this??
+// if (process.env.RESET_DB) {
+//   console.log('Resetting database')
+//   const seedDatabase = async () => {
+//     await Thought.deleteMany({})
+
+//     thought.forEach(() => {
+//       new Thought().save()
+//     })
+//   }
+//   seedDatabase()
+// }
+
 // Defines the port the app will run on. Defaults to 8080, but can be 
 // overridden when starting the server. For example:
 //
 //   PORT=9000 npm start
 const port = process.env.PORT || 8080
 const app = express()
 
-// Add middlewares to enable cors and json body parsing
+// MIDDLEWARES
 app.use(cors())
 app.use(bodyParser.json())
+app.use((req, res, next) => {
+  if (mongoose.connection.readyState === 1) {
+    next()
+  } else {
+    res.status(503).json({ error: 'Service unavailabale' })
+  }
+})
+
+app.get('/', async (req, res) => {
+  const { sort, page, tag, name } = req.query
+  const tagRegex = new RegExp(tag, "i")
+  const nameRegex = new RegExp(name, "i")
+
+  //Checks the sortquery, and sorts according to increasing date or decreasing likes
+  const buildSortQuery = (sort) => {
+    if (sort === "oldest") {
+      return { createdAt: 'asc' }
+    } else if (sort === "hearts") {
+      return { hearts: 'desc' }
+    } else {
+      return { createdAt: 'desc' }
+    }
+  }
 
-// Start defining your routes here
-app.get('/', (req, res) => {
-  res.send('Hello world')
+  //Checks how many results should be skipped
+  //e.g. if page = 1 it should skip none, if page it 2 it should skip 20
+  //because the limit it set to 20
+  const skipResults = (page) => {
+    return ((page - 1) * 20)
+  }
+
+  //Find thoughts based on name and tag, sort on date/likes,
+  //limit to 20 results per page, skip so that every page shows accurate results
+  let thoughts = await Thought.find()
+    .sort(buildSortQuery(sort))
+    .limit(20)
+    .skip(skipResults(page))
+    .exec()
+  if (tag) {
+    const filteredThoughts = await Thought.find({ 'tag': tagRegex })
+    res.json(filteredThoughts)
```

Comment:
Looks like you made progress with the tags idea.  Did you manage to get it working on the frontend?